Stabilizing multivariate normal approximation

han-ol opened the pull request:
This PR seeks to adress the stability of estimating mean and covariance using
```python
inference_network=bf.networks.PointInferenceNetwork(
        scores=dict(
            mvn=bf.scores.MultivariateNormalScore()
        )
)
# -> results in warning: MultivariateNormalScore is unstable.
```
It is fixed by using link function called PositiveDefinite, that is based on the Cholesky decomposition.

paul-buerkner commented:
Thanks! Is it intentional that this is to be merged into the `build-deepset-without-call` branch instead of into main (or dev)?

han-ol commented:
Definitely not intentional. This was actually developed against the current dev branch, so I changed it accordingly.

The new `links.PositiveDefinite` relies on `fill_triangular_matrix`. Which fails for tensorflow backend with `pred must not be a Python bool`, without it being clear where such a pred variable is defined

han-ol commented:
Pair programming with Valentin was successful (thx!) and we smoothed out the flaky tests. All backends support estimation of multivariate normal parameters now.

I removed some rough edges, added tests, so hopefully the codecov/path tests pass too now.
 
The PR is ready for review as soon as tests passed.

vpratz commented on bayesflow/scores/scoring_rule.py:
```diff
@@ -29,6 +29,8 @@ def __init__(
         self.subnets_kwargs = subnets_kwargs or {}
         self.links = links or {}
 
+        self.not_transforming_like_vector = []
```
Please add a comment documenting what this should contain, so people who want to set this in a subclass know what to use

LarsKue replied:
+1, I would also like to know if we can handle this in a better way than a special attribute.

vpratz replied:
Do we want this to be variable within a class, i.e., does this have to be an instance variable? If not, a constant [class variable](https://docs.python.org/3/tutorial/classes.html#class-and-instance-variables) might be an option, like
``` python
class ScoringRule:
    #: This variable lists ... (the #: syntax should allow sphinx to parse this)
    NOT_TRANSFORMING_LIKE_VECTOR = tuple()  # use immutable type tuple instead of list
```

han-ol replied:
Good idea to use a class variable for this! Implemented it in [d87b0b9](https://github.com/bayesflow-org/bayesflow/pull/380/commits/d87b0b911b6dbab8790dd390ce237738c48c6b1f). 

I added documenting comments as well.

han-ol replied:
Regarding a better way of handling this in general:

There is a discussion in #304 about the long term plans for special estimators and how they play with adapter transformations.
In my opinion, what we have here suffices for now and is a reasonable first step.

LarsKue commented on bayesflow/links/positive_definite.py:
```diff
@@ -0,0 +1,47 @@
+import keras
+
+# import numpy as np
```
Please remove this

LarsKue commented on bayesflow/networks/point_inference_network.py:
```diff
@@ -132,7 +132,7 @@ def call(
         if xz is None and not self.built:
             raise ValueError("Cannot build inference network without inference variables.")
         if conditions is None:  # unconditional estimation uses a fixed input vector
-            conditions = keras.ops.convert_to_tensor([[1.0]], dtype=keras.ops.dtype(xz))
```
Under what circumstances does this break? It seems like a good choice to use the dtype

han-ol replied:
Previously it broke because `xz` can be None. The method `estimate()` passes `xz=None` to `call()`.

(A non-None `xz` is passed to `call()` during training and is necessary for building the heads correctly.)

Solution: We can fix the dtype if `xz` is available and use float3 as a fallback if `xz` is None:

```python
conditions = keras.ops.convert_to_tensor(
    [[1.0]], dtype=keras.ops.dtype(xz) if xz is not None else "float32"
)
```

LarsKue commented on bayesflow/approximators/point_approximator.py:
```diff
@@ -119,6 +119,7 @@ def sample(
     def _prepare_conditions(self, conditions: dict[str, np.ndarray], **kwargs) -> dict[str, Tensor]:
         """Adapts and converts the conditions to tensors."""
         conditions = self.adapter(conditions, strict=False, stage="inference", **kwargs)
+        conditions.pop("inference_variables", None)
```
We could add this function to the ContinuousApproximator, if it is identical between it and the Point Approximator

han-ol replied:
Yes! This and similar refactoring of the ContinuousApproximator is a good idea (but I would keep them out of this PR).
There is also the option of moving the conversion to tensor into the adapter. Possibly with an optional bool flag convert_to_tensor that is by default False.

LarsKue commented on bayesflow/approximators/point_approximator.py:
```diff
@@ -130,6 +131,12 @@ def _apply_inverse_adapter_to_estimates(
         for score_key, score_val in estimates.items():
             processed[score_key] = {}
             for head_key, estimate in score_val.items():
+                if head_key in self.inference_network.scores[score_key].not_transforming_like_vector:
```
This seems like a very specific attribute. Can you explain some more about it?

han-ol replied:
I am marking this conversation as resolved in favor of the one where we have a thread about the attribute.

LarsKue commented on bayesflow/scores/multivariate_normal_score.py:
```diff
@@ -97,6 +98,8 @@ def sample(self, batch_shape: Shape, mean: Tensor, covariance: Tensor) -> Tensor
             A tensor of shape (batch_size, num_samples, D) containing the generated samples.
         """
+        if len(batch_shape) == 1:
+            batch_shape = (1,) + batch_shape
```
This fails if the user passes a list type batch shape

han-ol replied:
Good catch! Should we just use a tuple conversion?
```python
batch_shape = (1,) + tuple(batch_shape)
```

LarsKue replied:
I think the tuple conversion is ok. Internally, we only use tuple shapes, so I would have to think some more about changing the type hint itself.

But of course we want users to be able to pass lists where possible (and not a significant burden).

LarsKue commented on bayesflow/scores/scoring_rule.py:
```diff
@@ -95,14 +97,14 @@ def get_link(self, key: str) -> keras.Layer:
             Name of head for which to request a link.
-        shape: Shape
+        output_shape: Shape
             The necessary shape for the point estimators.
```
Could improve this doctring now that the name is changed

LarsKue commented on tests/test_links/test_links.py:
```diff
@@ -59,16 +52,21 @@ def test_quantile_ordering(quantiles, unordered):
 
-    output = activation(random_matrix_batch)
+    # Check if output is invertible
+    np.linalg.inv(output)
```
Don't use numpy for Tensors. The workflow won't show it, but this fails on Torch CUDA environments.

Either use numpy everywhere, or use tensors and then convert to numpy before calling it, or use `keras.ops.inv`, whichever is most appropriate (depends on the input requirements of `positive_definite` here)

LarsKue reviewed (approved):
Code overall looks good. See individual comments. I don't know enough about the requirements of this task to give a definite approval, but I trust that @han-ol did their research on this 🙂
